workload/schemachanger: fixes to get the workload stable

fqazi opened the pull request:
This pull request contains a number of fixes to get the schemachanger workload more stable

1. Fixes for various error conditions, where the schema changer workload was either expecting or not expecting correct errors
2. Additional logging for the schema changer workload to root cause errors
3. Removal of bad validation logic inside the schemachanger, which could lead to failures


Release justification: Low risk since we are only enabling a test again, which will find existing issues not create new ones.

fqazi committed:
workload/schemachanger: primary key detection is broken

Previously, the primary key detection was checking
for the wrong name inside the schema changer workload.
Additionally, when creating an index and selecting columns
an error is generated if the primary columns added again.
To address this, this patch will fix the detection of
primary key columns and errors during index creation
due to duplicate columns.

Release note: None

fqazi committed:
worklaod/schemachanger: query does not correctly detect empty mutations

Previously, when no mutations existed a schema changer
query would fail. Since we would call JSON array length on a
null object. This change uses the COALESCE function to
replace the NULL value with an empty mutation array.

Release note: None

fqazi committed:
workload/schemachanger: block alter table locality when not in multiregion

Previously, if the database was not setup for multiregion,
we still executed statements only allowed for multiregion
database. To address this, this patch adds detection for
this condition and skips the operation.

Release note: None

fqazi committed:
workload/schemachanger: generated columns cannot have default values

Previously, the schemachanger would incorrectly attempt
to add default values on generate columns and see unexpected
errors. This patch addresses this by expecting the correct
error when this happens.

Release note: None

fqazi committed:
workload/schemachanger: add logging for schema queries

Previously, we did not log the schema introspection queries,
their inputs or their results. We also did not have a simple
way to see all the queries in a txn. This patch adds logging
for schema introspection queries, txn logs, and dumping them
out on errors to make failure diagnosis easier.

Release note: None

fqazi committed:
workload/schemachanger: allow primary columns in secondary indexes

Previously, we used to disallowed having columns from the
primary index inside any secondary index. This patch allows
queries to be generated with this pattern, since no errors
will be generated.

Release note: None

fqazi committed:
workload/schemachanger: avoid using void as type

Previously, the randomized schema changer could use
the void type in different scenarios where it wasn't
allowed. To address this, this patch removes the
type from random selection.

Release note: None

fqazi committed:
workload/schemachanger: clean up connection pools

Previously, we would observe leak related failures
once the workload finished because connection pools
were never cleaned up. This patch adds logic to cleanup
the connection pools on workload completion.

Release note: None

fqazi committed:
sql/jobs: incorrect validation for jobs inside descriptors

Previously, it was possible to observe validation
failures because a job inside a descriptor could
not be found within the jobs table. This could happen
because the job would exist inside the job cache, but
not inside the actual jobs table yet. To address this,
this patch will get the validation to check the job
cache for entries in a transaction. Additionally,
extend the crdb_internal tables to show these jobs.

Release note (bug fix): Intermittent validation failures
could be observed on schema objects, where a job ID was
detected as missing when validating objects in a transaction.

fqazi committed:
workload/schemachanger: block schema change operations during PK swap

Previously, we could generate schema change operations while
PK swaps were active, which are expected to error out. To address,
this patch will skip generating statements that interact with
tables doing PK swaps.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76532)
<!-- Reviewable:end -->

fqazi committed:
worklaod/schemachanger: re-enable schema changer workload

Previously, the schema changer workload was disabled,
since there were a number of broken scenarios. With
the fixes gathered we should be able to re-enable this
test again. This patch will re-enable this workload
again to catch issues.

Release note: None

postamar reviewed (commented):
My review is quite superficial as I'm not familiar with this workload yet. Nothing strikes me as odd except the removal of the job validation. This will require some explanations to convince me.

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 5 of 5 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 2 of 2 files at r9, 4 of 4 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-:-Mx4usCvByRyLWsHnBY8:bcnpcls)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[-- commits, line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4stXV33dC8OTlpmgI):*
I'll be honest, I don't really know what screening queries are. I mean, I can guess, and this all seems legit in light of that.
___
*[-- commits, line 56 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4tW2T2GvB3hnHHmdF):*
Should we distinguish between indexes using primary and secondary encodings instead? I'm not sure.
___
*[pkg/jobs/validate.go, line 43 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4ttveCn6HL_PSUACG:-Mx4ttveCn6HL_PSUACH:b-yz8gqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab56d18389f3883c32401535e4b2348cb71136ea/pkg/jobs/validate.go#L43)):*
> ```Go
> 			// It may be possible for a job inside the descriptor and the jobs table
> 			// to be out of synchronization, since the two aren't updated in a
> 			// transactional manner.
> ```

Wait, really? How is this possible?
___
*[pkg/workload/schemachange/error_screening.go, line 961 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4sJfKBiJN7L6qedx8:-Mx4sJfKBiJN7L6qedx9:be1xojp) ([raw file](https://github.com/cockroachdb/cockroach/blob/88d9be76656b1e977e9d3f1604ef1be7530bbe17/pkg/workload/schemachange/error_screening.go#L961)):*
> ```Go
> 		ctx,
> 		tx,
> 		`SELECT EXISTS (SELECT * FROM [SHOW REGIONS FROM DATABASE])`,
> ```

TIL about `SELECT EXISTS`, thanks!
___
*[pkg/workload/schemachange/operation_generator.go, line 1965 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4sRSf3O_o2ORKbJes:-Mx4sRSf3O_o2ORKbJet:b-2ug4yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/de630281a588b24c885ab31fc988cee401abed6a/pkg/workload/schemachange/operation_generator.go#L1965)):*
> ```Go
> 	if columnForDefault.generated {
> 		og.expectedExecErrors.add(pgcode.InvalidTableDefinition)
> 	}
> ```

Should this also include `ON UPDATE` expressions?

More likely, `ON UPDATE` is simply not supported here, in which case, we should probably add support for it, perhaps as a follow-up.
___
*[pkg/workload/schemachange/operation_generator.go, line 274 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4udVy0uFzMh0PcYCo:-Mx4udVy0uFzMh0PcYCp:by4whn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2fafdb8649a28b186ce3c51c16cf55d23722d70/pkg/workload/schemachange/operation_generator.go#L274)):*
> ```Go
> 		// Dead statement lets retry.
> 		if stmt == "" {
> 			continue
> ```

Should we sleep or something before retrying?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[-- commits, line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4stXV33dC8OTlpmgI):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'll be honest, I don't really know what screening queries are. I mean, I can guess, and this all seems legit in light of that.
</blockquote></details>

Maybe something like "schema introspection queries" is probably cleaner language.
___
*[pkg/workload/schemachange/error_screening.go, line 961 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4sJfKBiJN7L6qedx8:-Mx5x1QZ6bsm41sCsx5L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/88d9be76656b1e977e9d3f1604ef1be7530bbe17/pkg/workload/schemachange/error_screening.go#L961)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

TIL about `SELECT EXISTS`, thanks!
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/operation_generator.go, line 1965 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4sRSf3O_o2ORKbJes:-Mx5xJtY7VgFpLIfRIcE:b-qkb8bw) ([raw file](https://github.com/cockroachdb/cockroach/blob/de630281a588b24c885ab31fc988cee401abed6a/pkg/workload/schemachange/operation_generator.go#L1965)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should this also include `ON UPDATE` expressions?

More likely, `ON UPDATE` is simply not supported here, in which case, we should probably add support for it, perhaps as a follow-up.
</blockquote></details>

I'll add a TODO for this
___
*[pkg/workload/schemachange/operation_generator.go, line 274 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4udVy0uFzMh0PcYCo:-Mx5x2TT3iFxjnchaX9r:b-tfu1go) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2fafdb8649a28b186ce3c51c16cf55d23722d70/pkg/workload/schemachange/operation_generator.go#L274)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should we sleep or something before retrying?
</blockquote></details>

I think this should be okay, we probably will have an artificial delay on retries as we query table information again. It won't ever turn into a tight spinning loop.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/jobs/validate.go, line 43 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4ttveCn6HL_PSUACG:-Mx61q2y1Wow109g6cy9:b5k7rlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab56d18389f3883c32401535e4b2348cb71136ea/pkg/jobs/validate.go#L43)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Wait, really? How is this possible?
</blockquote></details>

We wait till the end of a transaction to update the jobs table, so the Job ID may exist inside the descriptor, but we won't update the jobs table until we commit. So, calling crdb_internal.invalid_objects inside a transaction can be problematic, where we will think the descriptors are invalid.

Maybe the right answer is to also check the jobs cache at the same time instead of just the table.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[-- commits, line 56 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx4tW2T2GvB3hnHHmdF):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should we distinguish between indexes using primary and secondary encodings instead? I'm not sure.
</blockquote></details>

I think from a general randomization thing we are implicitly covering both cases, but I guess you're thinking a more targeted approach would be better?


<!-- Sent from Reviewable.io -->

ajwerner commented:
I think exposing the not-yet-written jobs in the current session to crdb_internal.jobs is a good idea

ajwerner commented on pkg/sql/crdb_internal.go:
```diff
@@ -4458,6 +4487,32 @@ func (m marshaledJobMetadataMap) GetJobMetadata(
 	return md, nil
 }
 
+func getPayloadAndProgressFromJob(
```
nit: call it  `getPayloadAndProgressFromJobRecord`?

ajwerner replied:
Also, give it some commentary explaining why it exists?

ajwerner commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -59,14 +65,26 @@ func columnExistsOnTable(
 	err = tx.QueryRow(ctx, query, args...).Scan(&b)
+	if err == nil {
+		og.LogQueryResults(
+			fmt.Sprintf(
+				"%q %q",
+				query,
+				args,
+			),
```
nit: this one feels potentially overwrapped

ajwerner reviewed (approved):
LGTM mod the nits and request for testing.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/crdb_internal.go, line 777 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx6H7EVR7K1VdkkaBTC-r21-777:-Mxot0ou39EDvgv8VQ7B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e60a6f27d43ab419210dbb87a2837220784c86d/pkg/sql/crdb_internal.go#L777)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you add some logictests which exercise this functionality? 
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 4490 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-Mx6H7EVR7K1VdkkaBTC-r21-4490:-Mxot1Wx-9qO3W2mblNU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e60a6f27d43ab419210dbb87a2837220784c86d/pkg/sql/crdb_internal.go#L4490)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Also, give it some commentary explaining why it exists? 
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/operation_generator.go, line 1768 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-MvtlEsBmUp_9qZrVSZD-r21-1768:-Mxoy1f_0X904Y6DuZ4-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e60a6f27d43ab419210dbb87a2837220784c86d/pkg/workload/schemachange/operation_generator.go#L1768)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

for these cases should we bubble up some well known error and then catch it at a higher level? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r28, 2 of 2 files at r30, 2 of 4 files at r31, 1 of 2 files at r32.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76532#-:-Mxp1SGN8r_aBFdTIUHU:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->

fqazi commented:
@ajwerner TFTR!

fqazi commented:
bors r+

fqazi commented:
bors r+
